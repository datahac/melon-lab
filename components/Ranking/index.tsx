import React, { useState } from 'react';
import * as R from 'ramda';
import Composer from 'react-composer';
import { NetworkConsumer } from '+/components/NetworkContext';
import { FundManagerConsumer } from '+/components/FundManagerContext';
import Ranking from '~/components/Ranking';
import RankingQuery from './data/ranking';
import * as Tm from '@melonproject/token-math';

const filterRankings = R.curryN(2, (search, fund) => {
  return fund.name.toLocaleLowerCase().includes(search);
});

const availableOrdering = [
  {
    name: 'Highest rank',
    value: '+rank',
  },
  {
    name: 'Lowest rank',
    value: '-rank',
  },
  {
    name: 'Highest price',
    value: '-price',
  },
  {
    name: 'Lowest price',
    value: '+price',
  },
  {
    name: 'Newest',
    value: '-inception',
  },
  {
    name: 'Oldest',
    value: '+inception',
  },
];

const sortRankings = ordering => (a, b) => {
  if (ordering === '+rank') {
    return Tm.greaterThan(a.rank, b.rank) ? 1 : -1;
  }

  if (ordering === '-rank') {
    return Tm.greaterThan(b.rank, a.rank) ? 1 : -1;
  }

  if (ordering === '+price' || ordering === '-price') {
    const priceA = Tm.toAtomic(a.sharePrice);
    const priceB = Tm.toAtomic(b.sharePrice);
    const difference = parseFloat(Tm.subtract(priceB, priceA).toString());
    return ordering === '-price' ? difference : -difference;
  }

  if (ordering === '+inception') {
    return new Date(a.inception) > new Date(b.inception) ? 1 : -1;
  }

  if (ordering === '-inception') {
    return new Date(b.inception) > new Date(a.inception) ? 1 : -1;
  }

  return 0;
};

const RankingContainer = ({}) => {
  const [ordering, setOrdering] = useState('+rank');
  const [search, setSearch] = useState('');

  return (
    <Composer components={[<NetworkConsumer />, <FundManagerConsumer />, <RankingQuery />]}>
      {([network, managerProps, rankingProps]) => {
        const funds =
          (!rankingProps.loading &&
            ((rankingProps.data && rankingProps.data.rankings) || [])
              .slice()
              .filter(filterRankings(search.toLocaleLowerCase()))
              .sort(sortRankings(ordering))) ||
          [];

        return (
          <Ranking
            availableOrdering={availableOrdering}
            associatedFund={managerProps.fund}
            funds={funds}
            loading={rankingProps.loading}
            search={search}
            ordering={ordering}
            setSearch={search => setSearch(search)}
            setOrdering={order => setOrdering(order)}
          />
        );
      }}
    </Composer>
  );
};

export default RankingContainer;
