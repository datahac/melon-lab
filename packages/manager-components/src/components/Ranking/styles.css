.ranking__table-wrap {
  overflow-x: auto;
}

.ranking__loading {
  margin: 0 auto;
  width: 100px;
}

.ranking__funds {
  display: table;
  border-top: 1px solid var(--color-border);
  border-right: 1px solid var(--color-border);
  border-left: 1px solid var(--color-border);
}

.ranking__filters {
  display: flex;
  align-items: center;
  margin-bottom: var(--space-m);
  width: 100%;
  justify-content: space-between;
}

.ranking__search :global(input) {
  max-width: 120px;

  @media (--viewport-m) {
    max-width: 200px;
  }
}
