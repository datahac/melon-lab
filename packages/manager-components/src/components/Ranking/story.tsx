import { action } from '@storybook/addon-actions';
import { storiesOf } from '@storybook/react';
import React from 'react';
import Ranking from './index';

const data = {
  rankingList: [
    {
      name: 'Black Turtle',
      address: '0x5bBA9263Ab1eA26FF9c0FeE3619e7AAf7C79E02b',
      inception: '26. Jul 2018 10:48',
      sharePrice: '1.0081',
      rank: 1,
    },
    {
      name: 'Quants',
      address: '0xFfDd0ae2efcBDb95fbCd189Fbb194708B3a3B3F9',
      inception: '24. Jul 2018 22:04',
      sharePrice: '1.0000',
      rank: 2,
    },
    {
      name: 'Titanium',
      address: '0xF1e9868Dc7b02AD30358476380Bfcb77A6c2e74e',
      inception: '26. Jul 2018 17:38',
      sharePrice: '1.0000',
      rank: 3,
    },
    {
      name: 'SolidEarnings',
      address: '0xE28073701a7Ba04D30059507B8Ec11Fcc4FD034e',
      inception: '25. Jul 2018 08:26',
      sharePrice: '1.0000',
      rank: 4,
    },
    {
      name: 'JZ',
      address: '0xDB14bA0Ad1939ACE877262Dadb56d15Fe4966bd5',
      inception: '23. Jul 2018 12:43',
      sharePrice: '1.0000',
      rank: 5,
    },
  ],
  loading: false,
  usersFund: '0xce35Be39A76706f314B2B430e80DaCB77c886242',
  search: '',
  ordering: '+rank',
  goToFund: action('goToFund'),
  onFilterChange: action('onFilterChange'),
  setOrdering: action('setOrdering'),
};

storiesOf('Components|Ranking', module).add('Default', () => {
  return <Ranking {...data} />;
});
