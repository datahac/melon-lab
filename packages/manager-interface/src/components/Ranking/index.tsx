import { networks } from '@melonproject/melon.js';
import { compose, withState, withPropsOnChange } from 'recompose';
import Ranking from '~/components/Ranking';
import { greaterThan } from '~/utils/functionalBigNumber';
import * as R from 'ramda';
import RankingQuery from './data/ranking';

const filterRankings = R.curryN(2, (search, fund) => {
  return fund.name.toLocaleLowerCase().includes(search);
});

const mapRankings = R.curryN(2, (network, fund) => ({
  ...fund,
  inception: fund.inception,
  sharePrice: fund.sharePrice,
  reportUrl: `https://${
    network === networks.KOVAN ? 'melon' : 'olympiad'
  }-reporting.now.sh/report/${fund.address}`,
}));

const sortRankings = ordering => (a, b) => {
  if (ordering === '+rank') {
    return greaterThan(a.rank, b.rank) ? 1 : -1;
  }

  if (ordering === '-rank') {
    return greaterThan(b.rank, a.rank) ? 1 : -1;
  }

  if (ordering === '+price') {
    return greaterThan(a.sharePrice, b.sharePrice) ? 1 : -1;
  }

  if (ordering === '-price') {
    return greaterThan(b.sharePrice, a.sharePrice) ? 1 : -1;
  }

  if (ordering === '+inception') {
    return new Date(a.inception) > new Date(b.inception) ? 1 : -1;
  }

  if (ordering === '-inception') {
    return new Date(b.inception) > new Date(a.inception) ? 1 : -1;
  }

  return 0;
};

const withSearchAndSorting = withPropsOnChange(
  ['loading', 'funds', 'search', 'ordering', 'network'],
  props => ({
    funds:
      (!props.loading &&
        (props.funds || [])
          .slice()
          .filter(filterRankings(props.search.toLocaleLowerCase()))
          .map(mapRankings(props.network))
          .sort(sortRankings(props.ordering))) ||
      [],
  }),
);

const withRanking = BaseComponent => baseProps => (
  <RankingQuery>
    {rankingProps => (
      <BaseComponent
        associatedFund={baseProps.associatedFund && baseProps.associatedFund.address}
        setOrdering={order => baseProps.setOrdering({ variables: { order } })}
        setSearch={search => baseProps.setSearch({ variables: { search } })}
        ordering={baseProps.ordering}
        search={baseProps.search}
        funds={rankingProps.data && rankingProps.data.rankings}
        network={baseProps.network}
        loading={rankingProps.loading}
      />
    )}
  </RankingQuery>
);

const withOrdering = withState('ordering', 'setOrdering', '+rank');
const withSearch = withState('search', 'setSearch', '');

export default compose(
  withRanking,
  withSearch,
  withOrdering,
  withSearchAndSorting,
)(Ranking);
