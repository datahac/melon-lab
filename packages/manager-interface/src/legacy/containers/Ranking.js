import moment from 'moment';
import { connect } from 'react-redux';
import { lifecycle } from 'recompose';
import Ranking from '@melonproject/manager-components/components/Ranking';
import { actions } from '../actions/ranking';
import displayNumber from '../utils/displayNumber';
import { actions as routeActions } from '../actions/routes';
import { greaterThan } from '../utils/functionalBigNumber';
import { networks } from '@melonproject/melon.js';

const mapStateToProps = state => ({
  rankingList: state.ranking.rankingList
    .filter(fund =>
      fund.name
        .toLocaleLowerCase()
        .includes(state.ranking.search.toLocaleLowerCase()),
    )
    .map(fund => ({
      ...fund,
      inception: moment(fund.inception).format('D. MMM YYYY HH:mm'),
      sharePrice: displayNumber(fund.sharePrice.toString()),
      reportUrl: `https://${
        state.ethereum.network === networks.KOVAN ? 'melon' : 'olympiad'
      }-reporting.now.sh/report/${fund.address}`,
    }))
    .sort((a, b) => {
      const { ordering } = state.ranking;
      if (ordering === '+rank') {
        return greaterThan(a.rank, b.rank) ? 1 : -1;
      } else if (ordering === '-rank') {
        return greaterThan(b.rank, a.rank) ? 1 : -1;
      } else if (ordering === '+price') {
        return greaterThan(a.sharePrice, b.sharePrice) ? 1 : -1;
      } else if (ordering === '-price') {
        return greaterThan(b.sharePrice, a.sharePrice) ? 1 : -1;
      } else if (ordering === '+inception') {
        return new Date(a.inception) > new Date(b.inception) ? 1 : -1;
      } else if (ordering === '-inception') {
        return new Date(b.inception) > new Date(a.inception) ? 1 : -1;
      }
      return 0;
    }),
  search: state.ranking.search,
  usersFund: state.app.usersFund,
  getFundLinkAction: fundAddress => routeActions.fund(fundAddress),
  loading: state.ranking.loading,
  ordering: state.ranking.ordering,
});

const mapDispatchToProps = dispatch => ({
  getRanking: () => dispatch(actions.getRanking()),
  onFilterChange: e => {
    dispatch(actions.setSearch(e.target.value));
  },
  setOrdering: ordering => dispatch(actions.setOrdering(ordering)),
  goToFund: fundAddress => dispatch(routeActions.fund(fundAddress)),
});

const RankingWithLifecycle = lifecycle({
  componentDidMount() {
    this.props.getRanking();
  },
})(Ranking);

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(RankingWithLifecycle);
