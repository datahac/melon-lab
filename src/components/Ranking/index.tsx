import React from 'react';
import * as R from 'ramda';
import Composer from 'react-composer';
import { NetworkConsumer } from '+/components/NetworkContext';
import { FundManagerConsumer } from '+/components/FundManagerContext';
import Ranking from '~/components/Ranking';
import RankingQuery from './data/ranking';
import * as Tm from '@melonproject/token-math';

const filterRankings = R.curryN(2, (search, fund) => {
  return fund.name.toLocaleLowerCase().includes(search);
});

const availableOrdering = [
  {
    name: 'Highest rank',
    value: '+rank',
  },
  {
    name: 'Lowest rank',
    value: '-rank',
  },
  {
    name: 'Highest price',
    value: '-price',
  },
  {
    name: 'Lowest price',
    value: '+price',
  },
  {
    name: 'Newest',
    value: '-inception',
  },
  {
    name: 'Oldest',
    value: '+inception',
  },
];

const mapRankings = R.curryN(2, (network, fund) => ({
  ...fund,
  inception: fund.inception,
  sharePrice: fund.sharePrice,
  reportUrl: `https://${
    network === 'KOVAN' ? 'melon' : 'olympiad'
  }-reporting.now.sh/report/${fund.address}`,
}));

const sortRankings = ordering => (a, b) => {
  if (ordering === '+rank') {
    return Tm.greaterThan(a.rank, b.rank) ? 1 : -1;
  }

  if (ordering === '-rank') {
    return Tm.greaterThan(b.rank, a.rank) ? 1 : -1;
  }

  if (ordering === '+price') {
    return Tm.greaterThan(a.sharePrice, b.sharePrice) ? 1 : -1;
  }

  if (ordering === '-price') {
    return Tm.greaterThan(b.sharePrice, a.sharePrice) ? 1 : -1;
  }

  if (ordering === '+inception') {
    return new Date(a.inception) > new Date(b.inception) ? 1 : -1;
  }

  if (ordering === '-inception') {
    return new Date(b.inception) > new Date(a.inception) ? 1 : -1;
  }

  return 0;
};

export default class RankingContainer extends React.PureComponent {
  state = {
    ordering: '+rank',
    search: '',
  };

  setOrdering = ordering => {
    this.setState({
      ordering,
    });
  };

  setSearch = search => {
    this.setState({
      search,
    });
  };

  render() {
    return (
      <Composer
        components={[
          <NetworkConsumer />,
          <FundManagerConsumer />,
          <RankingQuery />,
        ]}
      >
        {([network, managerProps, rankingProps]) => {
          const funds =
            (!rankingProps.loading &&
              ((rankingProps.data && rankingProps.data.rankings) || [])
                .slice()
                .filter(filterRankings(this.state.search.toLocaleLowerCase()))
                .map(mapRankings(network.network))
                .sort(sortRankings(this.state.ordering))) ||
            [];

          return (
            <Ranking
              availableOrdering={availableOrdering}
              associatedFund={managerProps.fund}
              funds={funds}
              loading={rankingProps.loading}
              search={this.state.search}
              ordering={this.state.ordering}
              setSearch={search => this.setSearch(search)}
              setOrdering={order => this.setOrdering(order)}
            />
          );
        }}
      </Composer>
    );
  }
}
