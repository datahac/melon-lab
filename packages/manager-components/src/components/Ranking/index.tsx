import React, { StatelessComponent } from 'react';
import Highlight from 'react-highlighter';
import Button from '~/blocks/Button';
import Input from '~/blocks/Input';
import Spinner from '~/blocks/Spinner';
import {
  CellBody,
  CellHead,
  Row,
  Table,
  TableBody,
  TableHead,
} from '~/blocks/Table';

import styles from './styles.css';

export interface RankingProps {
  rankingList?: any;
  loading?: boolean;
  usersFund?: string;
  onFilterChange?: () => void;
  ordering?: string;
  setOrdering?: (field) => void;
  search?: string;
  goToFund: (address) => void;
}

export const Ranking: StatelessComponent<RankingProps> = ({
  rankingList,
  loading,
  usersFund,
  onFilterChange,
  ordering,
  setOrdering,
  search,
  goToFund,
}) => {
  const onOrdering = (e, field) => {
    if (setOrdering) {
      ordering === `+${field}`
        ? setOrdering(`-${field}`)
        : setOrdering(`+${field}`);
    }
  };

  const onFundClick = (e, address) => {
    if (goToFund) {
      goToFund(address);
    }
  };

  const isActiveCol = colName => {
    if (ordering && ordering.includes(colName)) {
      if (ordering.includes('+')) {
        return 'asc';
      } else {
        return 'desc';
      }
    }
    return false;
  };

  return (
    <div className="ranking">
      <style jsx>{styles}</style>
      <h3>Melon Funds Ranking</h3>
      {loading ? (
        <div className="ranking__loading">
          <Spinner icon />
        </div>
      ) : (
        <div className="ranking__table-wrap">
          <Table>
            <TableHead>
              <Row isHead={true}>
                <CellHead
                  headFor="rank"
                  onClick={onOrdering}
                  sorted={isActiveCol('rank')}
                >
                  #
                </CellHead>
                <CellHead>
                  <Input
                    name="search"
                    placeholder="Search fund"
                    onChange={onFilterChange && onFilterChange}
                  />
                </CellHead>
                <CellHead
                  headFor="price"
                  onClick={onOrdering}
                  textAlign="right"
                  sorted={isActiveCol('price')}
                >
                  Share price
                </CellHead>
                <CellHead
                  headFor="inception"
                  onClick={onOrdering}
                  textAlign="right"
                  sorted={isActiveCol('inception')}
                >
                  Inception Date
                </CellHead>
              </Row>
            </TableHead>
            <TableBody>
              {rankingList.length > 0 &&
                rankingList.map(fund => (
                  <Row
                    key={fund.address}
                    active={fund.address === usersFund && true}
                    size="large"
                  >
                    <CellBody>{fund.rank}</CellBody>
                    <CellBody>
                      <Button
                        style="clear"
                        onClick={onFundClick}
                        buttonValue={fund.address}
                      >
                        <Highlight search={search}>{fund.name}</Highlight>
                      </Button>
                    </CellBody>
                    <CellBody textAlign="right">{fund.sharePrice}</CellBody>
                    <CellBody textAlign="right">{fund.inception}</CellBody>
                  </Row>
                ))}
            </TableBody>
          </Table>
        </div>
      )}
    </div>
  );
};

export default Ranking;
