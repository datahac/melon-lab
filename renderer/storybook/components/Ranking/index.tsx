import React, { StatelessComponent } from 'react';
import Card from '~/blocks/Card';
import Dropdown from '~/blocks/Dropdown';
import Input from '~/blocks/Input';
import Spinner from '~/blocks/Spinner';
import Link from '~/link';

import styles from './styles.css';

export interface RankingProps {
  loading?: boolean;
  funds?: any;
  ordering?: string;
  search?: string;
  setSearch: (search: string) => void;
  setOrdering: (field: string) => void;
  associatedFund?: string;
  availableOrdering;
}

export const Ranking: StatelessComponent<RankingProps> = ({
  loading,
  funds,
  ordering,
  search,
  setSearch,
  setOrdering,
  associatedFund,
  availableOrdering,
}) => (
  <div className="ranking">
    <style jsx>{styles}</style>
    {loading ? (
      <div className="ranking__loading">
        <Spinner icon />
      </div>
    ) : (
      <div className="ranking__table-wrap">
        <div className="ranking__filters">
          <div className="ranking__search">
            <Input
              name="search"
              placeholder="Search fund"
              value={search}
              onChange={e => setSearch(e.target.value)}
            />
          </div>
          <div className="ranking__sort">
            <Dropdown
              name="sort"
              options={availableOrdering}
              label="Sort by"
              value={ordering}
              onChange={e => setOrdering(e.target.value)}
            />
          </div>
        </div>

        {funds.length > 0 ? (
          <div className="ranking__funds">
            {funds.map(fund => (
              <Link
                key={fund.address}
                href={`/manage?address=${fund.address}&base=MLN&quote=WETH`}
              >
                <Card isActive={fund.address === associatedFund} {...fund} />
              </Link>
            ))}
          </div>
        ) : (
          <div className="ranking__no-funds">No funds are available</div>
        )}
      </div>
    )}
  </div>
);

export default Ranking;
