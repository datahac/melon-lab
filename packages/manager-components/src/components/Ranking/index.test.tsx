import React from 'react';
import Ranking from './index';

const mockCallback = jest.fn();
const mockCallbackFund = jest.fn();
const data = {
  rankingList: [
    {
      name: 'Black Turtle',
      address: '0x5bBA9263Ab1eA26FF9c0FeE3619e7AAf7C79E02b',
      inception: '26. Jul 2018 10:48',
      sharePrice: '1.0081',
      rank: 1,
    },
    {
      name: 'Quants',
      address: '0xFfDd0ae2efcBDb95fbCd189Fbb194708B3a3B3F9',
      inception: '24. Jul 2018 22:04',
      sharePrice: '1.0000',
      rank: 2,
    },
    {
      name: 'Titanium',
      address: '0xF1e9868Dc7b02AD30358476380Bfcb77A6c2e74e',
      inception: '26. Jul 2018 17:38',
      sharePrice: '1.0000',
      rank: 3,
    },
  ],
  usersFund: '0x5bBA9263Ab1eA26FF9c0FeE3619e7AAf7C79E02b',
  search: '',
  ordering: '+rank',
  goToFund: mockCallbackFund,
  onFilterChange: () => null,
  setOrdering: mockCallback,
};

describe('Ranking', () => {
  const defaultElement = <Ranking {...data} />;
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(defaultElement);
  });

  it('should render correctly when loading', () => {
    wrapper.setProps({ loading: true });
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('onChange order event', () => {
    wrapper
      .find('Dropdown')
      .simulate('change', { target: undefined, value: '+rank' });
    expect(mockCallback.mock.calls.length).toBe(1);
    expect(mockCallback.mock.calls[0][0]).toBe('+rank');
    wrapper
      .find('Dropdown')
      .simulate('change', { target: undefined, value: '-rank' });
    expect(mockCallback.mock.calls.length).toBe(2);
    expect(mockCallback.mock.calls[1][0]).toBe('-rank');
  });
});
