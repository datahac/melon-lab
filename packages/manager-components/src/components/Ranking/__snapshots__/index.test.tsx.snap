// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Holdings should render correctly 1`] = `
<div
  className="jsx-__scopedHash ranking"
>
  <JSXStyle
    css="__scoped"
    styleId="__scopedHash"
  />
  <h3
    className="jsx-__scopedHash"
  >
    Melon Funds Ranking
  </h3>
  <div
    className="jsx-__scopedHash ranking__loading"
  >
    <Spinner
      icon={true}
    />
  </div>
</div>
`;
